### PR TITLE
Perfect Periodic BC

### PR DESCRIPTION
**Pull request by LeilaGhaffari**
This PR is aiming at adding support for perfect periodic BC.
I have added two meshes in [examples/cube/](https://github.com/LeilaGhaffari/palace-lg/tree/lgh/periodic-bc/examples/cube). `cube.mesh` is a regular cube and `cube_periodic.mesh` is periodic.

@sebastiangrimberg , are the boundary conditions specified [here](https://github.com/LeilaGhaffari/palace-lg/blob/dda0768bc40dc680d157eeaf3b33cafa8909299a/examples/cube/cube_pec.json#L41) and [here](https://github.com/LeilaGhaffari/palace-lg/blob/dda0768bc40dc680d157eeaf3b33cafa8909299a/examples/cube/cube_periodic.json#L40) making sense?

Edit: Just noticed [mfem/data/periodic-cube.mesh](https://github.com/mfem/mfem/blob/master/data/periodic-cube.mesh) and it seems to be different than the one I created. This one fails too though.

### CODE REVIEW

**Comment by LeilaGhaffari** [comment]
@hughcars , I don't seem to have necessary permissions to add reviewers but I think this is in good shape for review. I would still need to add support for lexographic elements but that shouldn't block the review. Thanks!

File: scripts/schema/config/boundaries.json [hunk 1]
```diff
@@ -219,6 +219,22 @@
...
+        "type": "object",
+        "additionalProperties": false,
+        "required": ["DonorAttributes", "ReceiverAttributes"],
+        "properties":
+        {
+          "DonorAttributes": { "$ref": "#/$defs/DonorAttributes" },
+          "ReceiverAttributes": { "$ref": "#/$defs/ReceiverAttributes" }
+        }
+      }
+    },
```
Comment (hughcars) [inline scripts/schema/config/boundaries.json:237]: I don't believe this is correct: running `./scripts/validate-config examples/cube/cube_periodic_xy_user.json` from the root folder gives an error. This should also be updated once the interface is in terms of pairs of attributes `[[1,2], [3,4], [5,6]]` for instance in the case of the cube.

File: palace/fem/libceed/restriction.cpp [hunk 2]
```diff
@@ -12,16 +12,131 @@ namespace palace::ceed
...
+  Loc1.Transf.mesh = fespace.GetMesh();
+  Loc1.Transf.ElementType = mfem::ElementTransformation::ELEMENT;
+  Loc1.Transform(face_el->GetNodes(), face_ir);
+  mfem::IsoparametricTransformation face_tr;
+  face_tr.ElementNo = e;
+  face_tr.ElementType = mfem::ElementTransformation::BDR_ELEMENT;
+  face_tr.mesh = fespace.GetMesh();
+  face_tr.Attribute = fespace.GetMesh()->GetBdrAttribute(e);
+  mfem::DenseMatrix &face_pm = face_tr.GetPointMat();  // dim x dof
+  face_tr.Reset();
```
[the remark this answers is not in the document]
Comment (LeilaGhaffari) [reply palace/fem/libceed/restriction.cpp]: True, I had copy/pasted from a function where `face_tr` is returned!

File: palace/utils/configfile.hpp [hunk 3]
```diff
@@ -455,6 +455,23 @@ struct LumpedPortBoundaryData : public internal::DataMap<LumpedPortData>
...
 
+struct PeriodicData
+{
+public:
+  // Vector defining the direction and distance for this periodic boundary condition.
+  std::vector<double> translation = {0.0, 0.0, 0.0};
+  // List of boundary donor attributes for this periodic boundary condition.
+  std::vector<int> donor_attributes = {};
+  // List of boundary receiver attributes for this periodic boundary condition.
+  std::vector<int> receiver_attributes = {};
```
[the remark this answers is not in the document]
Comment (LeilaGhaffari) [reply palace/utils/configfile.hpp:466]: Per our conversation today, I'll leave the design as is for now.

File: palace/utils/configfile.hpp [hunk 4]
```diff
@@ -455,6 +455,23 @@ struct LumpedPortBoundaryData : public internal::DataMap<LumpedPortData>
   void SetUp(json &boundaries);
 };
 
+struct PeriodicData
+{
+public:
+  // Vector defining the direction and distance for this periodic boundary condition.
+  std::vector<double> translation = {0.0, 0.0, 0.0};
```
[the remark this answers is not in the document]
Comment (LeilaGhaffari) [reply palace/utils/configfile.hpp]: I was thinking of populating the translation vector inside `LoadMesh()` when we add that functionality. In that case, Distance and Direction wouldn’t be required.

**Review by hughcars — CHANGES_REQUESTED** [review]
Almost done. The issue with the hex/prism/tet waveguide is going to mean the gmsh meshes need to be refined, cavity rebaselined and then the tables of the corresponding modes filled in. Additionally, do not need to remove the attributes on donor and receiver (and can patch mfem to not do the same as well) then we can add computing the power flux for each mode to the regression tests.

**Comment by hughcars** [comment]
I've addressed all the issues here in https://github.com/awslabs/palace/tree/hughcars/periodic-bc including merging main in. I'm not sure how to merge those commits onto this branch, given it's originally from a fork. You can either merge them in, or I can open another PR from that branch. It should just be a fast forward from this branch.

**Comment by LeilaGhaffari** [comment]
Thanks @hughcars for fixing the issues. I just pushed your fixes to this branch. Is there anything else I can help with before merging?

**Comment by hughcars** [comment]
> Thanks @hughcars for fixing the issues. I just pushed your fixes to this branch. Is there anything else I can help with before merging?

I just read through it again and found a few small things I missed, there's one more commit on `hughcars/periodic-bc` then we're good to merge.

**Comment by LeilaGhaffari** [comment]
> > Thanks @hughcars for fixing the issues. I just pushed your fixes to this branch. Is there anything else I can help with before merging?
> 
> I just read through it again and found a few small things I missed, there's one more commit on `hughcars/periodic-bc` then we're good to merge.

Just pushed it, thanks!

**Review by hughcars — APPROVED** [review]

**Pull request merged**